// passport-stormpath/lib/strategy.js
var stormpath = require('stormpath');

function lookup(obj, field) {
    if (!obj) { return null; }

    if(typeof(obj) === 'object'){
        for(var prop in obj) {
            if(typeof(obj[prop]) === 'object'){
                return lookup(obj[prop],field);
            }
        }
    }

    var chain = field.split(']').join('').split('[');
    for (var i = 0, len = chain.length; i < len; i++) {
        var prop = obj[chain[i]];
        if (typeof(prop) === 'undefined') { return null; }
        if (typeof(prop) !== 'object') { return prop; }
        obj = prop;
    }

    return null;
}

function lookupRec(object, field){
    if(!object){
        return null;
    }

    for(var prop in object) {
        // propertyName is what you want
        // you can get the value like this: myObject[propertyName]
        if(typeof(object[prop]) === 'object'){
            return lookup(object[prop],field);
        }        
    }
}

function Strategy(o){
    var opts = o || {};

    this._usernameField = opts.usernameField || 'username';
    this._passwordField = opts.passwordField || 'password';

    var apiKeyId = opts.apiKeyId || process.env['STORMPATH_API_KEY_ID'] || "";
    var apiKeySecret = opts.apiKeySecret || process.env['STORMPATH_API_KEY_SECRET'] || "";
    var appHref = opts.appHref || process.env['STORMPATH_APP_HREF'] || "";

    var self = this;

    if(opts.spClient){
        self.spClient = opts.spClient;
    }else{
        self.spClient = new stormpath.Client({
            apiKey: new stormpath.ApiKey(apiKeyId,apiKeySecret)
        });
    }

    if(opts.spApp){
        self.spApp = opts.spApp;
    }else{
        self.spClient.getApplication(appHref,
            function(err,app){
                self.spApp = app;
                if(err){
                    throw err;
                }
            }
        );
    }
    self.serializeUser = function(user, done) {
        done(null, user.href);
    };

    self.deserializeUser = function(userHref, done) {
        self.spClient.getAccount(userHref,function(err,account){
            done(err,account);
        });
    };
    return this;
}

Strategy.prototype.name = "stormpath";

Strategy.prototype.authenticate = function(req, options) {
    //
    //TODO add recursive parsing through the body data may be wrapped in an object
    //
    options = options || {};
    var self = this;

    var username = lookup(req.body, this._usernameField) || lookup(req.query, this._usernameField);
    var password = lookup(req.body, this._passwordField) || lookup(req.query, this._passwordField);
    var data = {username:username,password:password};

    if (!username || !password) {
        return self.fail({ message: options.badRequestMessage || 'Missing credentials' }, 400);
    }

    self.spApp.authenticateAccount(data, function onAuthcResult(err, result) {
        if (err) {
            return self.fail({message:err.userMessage},400);
        }else{
            self.success(result.account);
        }
    });
};

module.exports = Strategy;
